### PR TITLE
rgw_sal_motr: [CORTX-31196]PutBucketTagging: Call put_info() with false to update the attributes

### PR DESCRIPTION
**Pull request by selvakumaar5496**
Signed-off-by: Selvakumar <selva.k.nambi@seagate.com>

Problem:
Enabled PutBucketTagging API by removing negation in exclusive flag(#172 ). PutBucketVersioning API is affected due to this change. 

1. CreateBucket initially calls put_info() where exclusive=true, hence update=False.
2. PutBucketTagging calls put_info() where exclusive=false to make update=True. If PutBucketTagging is called with exclusive=True, it will try to CreateBucket again as mentioned in step 1

Resolution:
By setting exclusive to false in put_info() function updates the attributes and  resolves the issue

Results:
```
[root@ssc-vm-rhev4-2587 build]# aws --endpoint=$endpoint_url s3api put-bucket-versioning --bucket test-buckets --versioning-configura Status=Suspended
[root@ssc-vm-rhev4-2587 build]# aws --endpoint=$endpoint_url s3api get-bucket-versioning --bucket test-buckets
{
    "Status": "Suspended",
    "MFADelete": "Disabled"
}
[root@ssc-vm-rhev4-2587 build]# aws --endpoint=$endpoint_url s3api put-bucket-versioning --bucket test-buckets --versioning-configura Status=Enabled
[root@ssc-vm-rhev4-2587 build]# aws --endpoint=$endpoint_url s3api get-bucket-versioning --bucket test-buckets                       {
    "Status": "Enabled",
    "MFADelete": "Disabled"
}
[root@ssc-vm-rhev4-2587 build]#
[root@ssc-vm-rhev4-2587 build]#
[root@ssc-vm-rhev4-2587 build]#
[root@ssc-vm-rhev4-2587 build]#
[root@ssc-vm-rhev4-2587 build]# aws --endpoint=$endpoint_url s3api put-bucket-tagging --bucket test-buckets --tagging '{"TagSet":[{ ": "team", "Value": "c5-solvers12345" }]}'
[root@ssc-vm-rhev4-2587 build]# aws --endpoint=$endpoint_url s3api get-bucket-tagging --bucket test-buckets                          {
    "TagSet": [
        {
            "Key": "team",
            "Value": "c5-solvers12345"
        }
    ]
}
```

```
[root@ssc-vm-rhev4-2587 build]# bin/radosgw-admin quota set --uid mgwuser --bucket test-buckets --max-size 1024                      2022-05-10T05:21:19.222-0600 7fcb064db5c0 -1 WARNING: all dangerous and experimental features are enabled.
2022-05-10T05:21:19.263-0600 7fcb064db5c0 -1 WARNING: all dangerous and experimental features are enabled.
2022-05-10T05:21:19.265-0600 7fcb064db5c0 -1 WARNING: all dangerous and experimental features are enabled.
2022-05-10T05:21:19.265-0600 7fcb064db5c0  1 RGW CONF BACKEND STORE = motr
2022-05-10T05:21:19.265-0600 7fcb064db5c0  1 RGW BACKEND STORE = motr
2022-05-10T05:21:19.265-0600 7fcb064db5c0  0 INFO: motr my endpoint: 192.168.180.182@tcp:12345:4:1
2022-05-10T05:21:19.265-0600 7fcb064db5c0  0 INFO: motr ha endpoint: inet:tcp:10.230.241.85@22001
2022-05-10T05:21:19.265-0600 7fcb064db5c0  0 INFO: motr my fid:      0x7200000000000001:0x0
2022-05-10T05:21:19.265-0600 7fcb064db5c0  0 INFO: motr profile fid: 0x7000000000000001:0x51
2022-05-10T05:21:19.265-0600 7fcb064db5c0  0 INFO: init flags:       0
2022-05-10T05:21:19.265-0600 7fcb064db5c0  0 INFO: motr admin endpoint: inet:tcp:10.230.241.85@22502
2022-05-10T05:21:19.265-0600 7fcb064db5c0  0 INFO: motr admin fid:   0x7200000000000001:0x2e
motr[381283]:  ae50  ERROR  [dix/req.c:800:dix_idxop_meta_update_ast_cb]  All items are failed
motr[381283]:  ae50  ERROR  [dix/req.c:800:dix_idxop_meta_update_ast_cb]  All items are failed
motr[381283]:  ae50  ERROR  [dix/req.c:800:dix_idxop_meta_update_ast_cb]  All items are failed
motr[381283]:  5e50  ERROR  [dix/req.c:800:dix_idxop_meta_update_ast_cb]  All items are failed
motr[381283]:  ae50  ERROR  [dix/req.c:800:dix_idxop_meta_update_ast_cb]  All items are failed
2022-05-10T05:21:20.572-0600 7fcb064db5c0  0 Cache miss: name = mgwuser, rc = -2
2022-05-10T05:21:20.578-0600 7fcb064db5c0  0 Put into cache: name = mgwuser
2022-05-10T05:21:20.579-0600 7fcb064db5c0  0 Cache miss: name = test-buckets, rc = -2
2022-05-10T05:21:20.583-0600 7fcb064db5c0  0 Put into cache: name = test-buckets
2022-05-10T05:21:20.596-0600 7fcb064db5c0  0 Put into cache: name = test-buckets
[root@ssc-vm-rhev4-2587 build]#
[root@ssc-vm-rhev4-2587 build]# bin/radosgw-admin bucket stats --uid mgwuser --bucket test-buckets                                   2022-05-10T05:21:36.527-0600 7f90c8c835c0 -1 WARNING: all dangerous and experimental features are enabled.
2022-05-10T05:21:36.545-0600 7f90c8c835c0 -1 WARNING: all dangerous and experimental features are enabled.
2022-05-10T05:21:36.546-0600 7f90c8c835c0 -1 WARNING: all dangerous and experimental features are enabled.
2022-05-10T05:21:36.546-0600 7f90c8c835c0  1 RGW CONF BACKEND STORE = motr
2022-05-10T05:21:36.546-0600 7f90c8c835c0  1 RGW BACKEND STORE = motr
2022-05-10T05:21:36.546-0600 7f90c8c835c0  0 INFO: motr my endpoint: 192.168.180.182@tcp:12345:4:1
2022-05-10T05:21:36.546-0600 7f90c8c835c0  0 INFO: motr ha endpoint: inet:tcp:10.230.241.85@22001
2022-05-10T05:21:36.546-0600 7f90c8c835c0  0 INFO: motr my fid:      0x7200000000000001:0x0
2022-05-10T05:21:36.546-0600 7f90c8c835c0  0 INFO: motr profile fid: 0x7000000000000001:0x51
2022-05-10T05:21:36.546-0600 7f90c8c835c0  0 INFO: init flags:       0
2022-05-10T05:21:36.546-0600 7f90c8c835c0  0 INFO: motr admin endpoint: inet:tcp:10.230.241.85@22502
2022-05-10T05:21:36.546-0600 7f90c8c835c0  0 INFO: motr admin fid:   0x7200000000000001:0x2e
motr[381648]:  9e50  ERROR  [dix/req.c:800:dix_idxop_meta_update_ast_cb]  All items are failed
motr[381648]:  9e50  ERROR  [dix/req.c:800:dix_idxop_meta_update_ast_cb]  All items are failed
motr[381648]:  9e50  ERROR  [dix/req.c:800:dix_idxop_meta_update_ast_cb]  All items are failed
motr[381648]:  6e50  ERROR  [dix/req.c:800:dix_idxop_meta_update_ast_cb]  All items are failed
motr[381648]:  9e50  ERROR  [dix/req.c:800:dix_idxop_meta_update_ast_cb]  All items are failed
2022-05-10T05:21:37.551-0600 7f90c8c835c0  0 Cache miss: name = mgwuser, rc = -2
2022-05-10T05:21:37.565-0600 7f90c8c835c0  0 Put into cache: name = mgwuser
2022-05-10T05:21:37.565-0600 7f90c8c835c0  0 Cache miss: name = test-buckets, rc = -2
2022-05-10T05:21:37.575-0600 7f90c8c835c0  0 Put into cache: name = test-buckets
2022-05-10T05:21:37.575-0600 7f90c8c835c0  0 Cache miss: name = mgwuser, rc = -2
2022-05-10T05:21:37.581-0600 7f90c8c835c0  0 Put into cache: name = mgwuser
2022-05-10T05:21:37.587-0600 7f90c8c835c0  0 Cache miss: name = test-buckets, rc = -2
[
    {
        "bucket": "test-buckets",
        "num_shards": 1,
        "tenant": "",
        "zonegroup": "0956b174-fe14-4f97-8b50-bb7ec5e1cf62",
        "placement_rule": "default",
        "explicit_placement": {
            "data_pool": "",
            "data_extra_pool": "",
            "index_pool": ""
        },
        "id": "",
        "marker": "",
        "index_type": "Normal",
        "owner": "mgwuser",
        "ver": "",
        "master_ver": "",
        "mtime": "0.000000",
        "creation_time": "0.000000",
        "max_marker": "",
        "usage": {},
        "bucket_quota": {
            "enabled": true,
            "check_on_raw": false,
            "max_size": 1024,
            "max_size_kb": 1,
            "max_objects": -1
        },
        "tagset": {
            "team": "c5-solvers12345"
        }
    }
]
2022-05-10T05:21:37.599-0600 7f90c8c835c0  0 Put into cache: name = test-buckets
```

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [x] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/rgw/rgw_sal_motr.cc [hunk 1]
```diff
@@ -721,6 +721,9 @@ int MotrBucket::put_info(const DoutPrefixProvider *dpp, bool exclusive, ceph::re
   bufferlist bl;
   struct MotrBucketInfo mbinfo;
   string tenant_bkt_name = get_bucket_name(info.bucket.tenant, info.bucket.name);
+  
+  // Set exclusive to false for put bucket tags
+  exclusive = (attrs.find(RGW_ATTR_TAGS) == attrs.end());
```
Comment (cdeshmukh) [inline src/rgw/rgw_sal_motr.cc:726]: This change should not affect any other API, except PutBucketTagging, please test enough.
Does this come in any Obj APIs too where Tagging are being added for an Object.

Comment (selvakumaar5496) [reply src/rgw/rgw_sal_motr.cc:726]: 1. API's with known issues are not blocked now and works fine.
2. No merge_and_store_attrs is not used by any Obj Tagging APIs

**Review by sjain09 — APPROVED** [review]

File: src/rgw/rgw_sal_motr.cc (hunk 1 above)
Comment (cdeshmukh) [reply src/rgw/rgw_sal_motr.cc:726]: Ok fine.

**Comment by cdeshmukh** [comment]
rebase plz

File: src/rgw/rgw_sal_motr.cc (hunk 1 above)
Comment (jjxsg) [inline src/rgw/rgw_sal_motr.cc:726]: @selvakumaar5496 can you please confirm if this agreed solution or temporary workaround?
If it's a workaround, please add a TODO.

Comment (cdeshmukh) [reply src/rgw/rgw_sal_motr.cc:726]: @jjxsg its discussed solution, plus tested with different APIs to confirm any possible issues, please check in [PR description](https://github.com/Seagate/cortx-rgw/pull/199#issue-1231077520).
exclusive=false needed to make update=true @ line 737, to update existing bucket entry for tags update. 

Same way different attributes seen detected, like seen determining for RGW_ATTR_ACL in rgw_op_get_bucket_policy_from_attr() in rgw_sal_rados.cc, and likewise for other attributes across the code.

Comment (selvakumaar5496) [reply src/rgw/rgw_sal_motr.cc]: I have updated the PR as discussed with @sumedhak27 and @sachinpunadikar . We can consider this as a solution if sachin and sumedh agrees
cc: @jjxsg

**Review by sumedhak27 — APPROVED** [review]

File: src/rgw/rgw_sal_motr.cc [hunk 2]
```diff
@@ -275,7 +275,7 @@ int MotrUser::create_bucket(const DoutPrefixProvider* dpp,
     // Create a new bucket: (1) Add a key/value pair in the
     // bucket instance index. (2) Create a new bucket index.
     MotrBucket* mbucket = static_cast<MotrBucket*>(bucket.get());
-    ret = mbucket->put_info(dpp, y, ceph::real_time())? :
+    ret = mbucket->put_info(dpp, true, ceph::real_time())? :
```
Comment (sachinpunadikar) [inline src/rgw/rgw_sal_motr.cc:278]: Add comment as below:
"put_info" accepts boolean value mentioning whether to create new or update existing. "yield" is not a boolean flag, hence explicitly passing true, meaning create a new record.

File: src/rgw/rgw_sal_motr.cc [hunk 3]
```diff
@@ -898,7 +898,8 @@ int MotrBucket::merge_and_store_attrs(const DoutPrefixProvider *dpp, Attrs& new_
   for (auto& it : new_attrs)
     attrs[it.first] = it.second;
 
-  return put_info(dpp, y, ceph::real_time());
+  // Call put_info with false to update the attributes
```
Comment (sachinpunadikar) [inline src/rgw/rgw_sal_motr.cc:901]: Update the comment as :
"merge_and_store_attrs" meant for updating existing metadata, hence explicitly passing false. Passing "yield" was incorrect

**Review by sachinpunadikar — CHANGES_REQUESTED** [review]
work on comments

File: src/rgw/rgw_sal_motr.cc (hunk 2 above)
Comment (selvakumaar5496) [reply src/rgw/rgw_sal_motr.cc:281]: Done

File: src/rgw/rgw_sal_motr.cc (hunk 1 above)
Comment (jjxsg) [reply src/rgw/rgw_sal_motr.cc]: Thanks for clarification @cdeshmukh @selvakumaar5496

File: src/rgw/rgw_sal_motr.cc (hunk 3 above)
Comment (selvakumaar5496) [reply src/rgw/rgw_sal_motr.cc]: Done

**Review by jjxsg — APPROVED** [review]

File: src/rgw/rgw_sal_motr.cc [hunk 4]
```diff
@@ -275,7 +275,10 @@ int MotrUser::create_bucket(const DoutPrefixProvider* dpp,
     // Create a new bucket: (1) Add a key/value pair in the
     // bucket instance index. (2) Create a new bucket index.
     MotrBucket* mbucket = static_cast<MotrBucket*>(bucket.get());
-    ret = mbucket->put_info(dpp, y, ceph::real_time())? :
+	
+	// "put_info" accepts boolean value mentioning whether to create new or update existing.
```
Comment (cdeshmukh) [inline src/rgw/rgw_sal_motr.cc:279]: correct indentation

**Review by sachinpunadikar — APPROVED** [review]
+1

File: src/rgw/rgw_sal_motr.cc [hunk 5]
```diff
@@ -898,7 +901,9 @@ int MotrBucket::merge_and_store_attrs(const DoutPrefixProvider *dpp, Attrs& new_
   for (auto& it : new_attrs)
     attrs[it.first] = it.second;
 
-  return put_info(dpp, y, ceph::real_time());
+  // "merge_and_store_attrs" is meant to updating existing metadata,
```
Comment (cdeshmukh) [inline src/rgw/rgw_sal_motr.cc:904]: why merge_and_store_attrs() needs to be mentioned here?

File: src/rgw/rgw_sal_motr.cc [hunk 6]
```diff
@@ -898,7 +901,9 @@ int MotrBucket::merge_and_store_attrs(const DoutPrefixProvider *dpp, Attrs& new_
   for (auto& it : new_attrs)
     attrs[it.first] = it.second;
 
-  return put_info(dpp, y, ceph::real_time());
+  // "merge_and_store_attrs" is meant to updating existing metadata, 
+  // hence explicitly passing false. Passing "yield" was incorrect
+  return put_info(dpp, false, ceph::real_time());
```
Comment (cdeshmukh) [inline src/rgw/rgw_sal_motr.cc:906]: so only during create bucket, this bool is set to false, so update route is ruled out. and then any operation on bucket, like elabling versioning, tagging, set quota, it should be false(default), so bucket attribute update is allowed for such APIs!

File: src/rgw/rgw_sal_motr.cc (hunk 3 above)
Comment (cdeshmukh) [reply src/rgw/rgw_sal_motr.cc]: @sachinpunadikar doesnt look right, why we need to mention merge_and_store_attrs() here ?

File: src/rgw/rgw_sal_motr.cc (hunk 5 above)
Comment (selvakumaar5496) [reply src/rgw/rgw_sal_motr.cc]: Modifiedit to put_info

File: src/rgw/rgw_sal_motr.cc (hunk 4 above)
Comment (selvakumaar5496) [reply src/rgw/rgw_sal_motr.cc]: Done

File: src/rgw/rgw_sal_motr.cc [hunk 7]
```diff
@@ -898,7 +900,9 @@ int MotrBucket::merge_and_store_attrs(const DoutPrefixProvider *dpp, Attrs& new_
   for (auto& it : new_attrs)
     attrs[it.first] = it.second;
 
-  return put_info(dpp, y, ceph::real_time());
+  // "put_info" is meant to update existing metadata, by
+  // explicitly passing false. Passing "yield" was incorrect
+  return put_info(dpp, false, ceph::real_time());
```
Comment (cdeshmukh) [inline src/rgw/rgw_sal_motr.cc:905]: // "put_info" second bool argument is meant to update existing metadata, which is not needed here, 
// so explicitly passing false.

No need to mention about yield here

Comment (selvakumaar5496) [reply src/rgw/rgw_sal_motr.cc:905]: Done

**Review by cdeshmukh — APPROVED** [review]

**Pull request merged**